### ISSUE TITLE
Carefull with reading location table from CALCPNT.HIS

### ISSUE THREAD

**Issue by dquang** [HKV-products-services/hkvsobekpy#1]
Hi,

thank you very much for the package. I have been working also with SOBEK 2 and was writing a similar function to read .HIS file in R. I found out that there was a small (might be a big) problem in the binary file with the location table. Everything went well with the table structure of an integer + a character vector (4 + 20 bytes) until I regconized that the result was wrong while reading a CALCPNT.HIS. From the reading, instead of 3 (for 3rd Location), it was 4, and there were some other errors with location indexes bigger than the total of locations. I tried it with different CALCPNT.HIS files from different models (different river networks) and it was always wrong.

I though that it was something wrong in R or with my code and I tried with other languages included your piece of code in this package but it came with the same result. Then I read the source code of the ReaHis.exe (come with Sobek, written by Deltares) and found out that they did not export the location table with the read integer, but with a normal index in a loop. Therefore, I suggest that you should change this location id to  a normal index instead of the read integer.

Regards,
Quang

**Comment by mattijn** [comment]
Thanks for raising the issue! I'm trying to understand what you're saying, but not sure if I get it. 

Could you maybe provide the CALCPNT.HIS you are working with and exactly describe which lines of code you have executed and describe where you see different results compare what you expect to see based on the official documentation?

For comparison I did the following to compare HKVSOBEKPY results with ODS VIEW results and they give the same results:
```python
import hkvsobekpy as hkv

his_file = r'path\to\CALCPNT.HIS'
his_object = hkv.read_his.ReadMetadata(his_file, hia_file='none')

locs = his_object.GetLocations()
pars = his_object.GetParameters()

df = his_object.DataFrame()

print([pars[0], locs[0]])
['Waterlevel mean (m A', 'N_NDB_0']
```
```python
df[pars[0]][locs[0]].plot()
```
![download](https://user-images.githubusercontent.com/5186265/45298772-d359a200-b509-11e8-8d45-49bdbc30c7ef.png)

And using the same parameter + location (`Waterlevel mean (m A` + `N_NDB_0`) I see the following results in ODS_View 3.23:
![download2](https://user-images.githubusercontent.com/5186265/45298879-1f0c4b80-b50a-11e8-93bf-813c955cfff5.png)

**Comment by mattijn** [comment]
Also which version of HKVSOBEKPY are you using? Make sure you are using `1.3.3` as in `1.3.2` there was a bug in the function `his_object.DataFrame()` (but that seems not related to the behavior you are describing)

**Comment by dquang** [comment]
thanks for quick reply! I am talking about the lines from 225-227 in the hkvsobekpy/his.py:

`for i in range(self.hisFile.locationInfo.numLoc):            
   self.hisFile.locationInfo.id.append(np.fromfile(f,np.int,1)[0])
   self.hisFile.locationInfo.locations.append(f.read(20).decode('windows-1252').rstrip())`

the reading of location id with np.fromfile(f, np.int, 1)[0] is ok with other .his (ex. reachseg.his) but is not with calcpnt.his. Problem is, it is not wrong for all, but some. Could you please try to check again while reading calcpnt.his if the max of location ids is bigger than the total location - that should not be the case? Because the data matrix that comes at the end of the file is arranged with the location id. If we have location id that is bigger than the total location, the matrix indexing will be out of bound!

Unfortunately, I cannot provide the calcpnt.his without having an official permission. We are using Sobek Version 2.13.002.

**Comment by mattijn** [comment]
We have test it with a calcpnt.his from Sobek Version 2.13.002, but we could unfortunately not reproduce the issue you are raising. 
I did notice that the `locationInfo.id` contains multiple zeros and is not always ordered, but since we use the `locationInfo.locations` this is fine. Furthermore `locationInfo.numLocs` is equal to `len(locationInfo.locations)`.

**Comment by dquang** [comment]
I have installed the full hkvsobekpy package and test again. Everything is fine! Sorry, it was my fault thinking that you use the location.id as the index to get the data from the matrix at the end of the .his file. 

So you use the locations.index to locate the right column for the data of a "node". That makes sense, since the source code of reahis.exe also use the same thing, and the location table is simply sorted and the location.id written in the .his file is useless. We should ask Deltares why they write an "int" to the matrix and do not use it :)

I learned from your code that there is a .hia file storing "long names" of the nodes.

Again, thank you very much!
